pynamodb/connection/base.py

```diff
@@ -338,35 +338,55 @@ def _make_api_call(self, operation_name, operation_kwargs):
         1. It's faster to avoid using botocore's response parsing
         2. It provides a place to monkey patch requests for unit testing
         """
+
         operation_model = self.client._service_model.operation_model(operation_name)
         request_dict = self.client._convert_to_request_dict(
             operation_kwargs,
-            operation_model
+            operation_model,
         )
+
+        if self._extra_headers is not None:
+            request_dict['headers'].update(self._extra_headers)
+
         prepared_request = self._create_prepared_request(request_dict, operation_model)
 
         for i in range(0, self._max_retry_attempts_exception + 1):
             attempt_number = i + 1
             is_last_attempt_for_exceptions = i == self._max_retry_attempts_exception
 
-            response = None
+            http_response = None
             try:
-                proxies = getattr(self.client._endpoint, "proxies", None)
-                # After the version 1.11.0 of botocore this field is no longer available here
-                if proxies is None:
-                    proxies = self.client._endpoint.http_session._proxy_config._proxies
-
-                response = self.requests_session.send(
-                    prepared_request,
-                    timeout=self._request_timeout_seconds,
-                    proxies=proxies,
-                )
-                data = response.json()
-            except (requests.RequestException, ValueError) as e:
+                if i > 0:
+                    # If there is a stream associated with the request, we need
+                    # to reset it before attempting to send the request again.
+                    # This will ensure that we resend the entire contents of the
+                    # body.
+                    prepared_request.reset_stream()
+
+                    # Create a new request when retried (including a new signature).
+                    prepared_request = self._create_prepared_request(request_dict, operation_model)
+
+                # Implement the before-send event from botocore
+                event_name = 'before-send.dynamodb.%s' % operation_model.name
```

Comment:
I think percent-formatting is deprecated now?

Reply:
Yep, took this from botocore which supports 2.6 still. I'll switch to `.format` now that we've dropped support for that.